### ISSUE TITLE
Respect path mapping in all cases

### ISSUE THREAD

**Issue by wessberg** [wessberg/rollup-plugin-ts#14]
Currently, path mapping is supported and the interop with Rollup works fine with the `resolveId` hook. However, it only works for files that has been added to the language service which may not be the case for external dependencies with `.js` extensions with `allowJs` is false. The consequence is that authoring reports no errors with path mapping, but then bundling with Rollup fails to resolve the mapped paths.

**Comment by Septh** [comment]
Just met this very issue :)

I have a `tsconfig.json` file for my project with the `allowJs` and `checkJs` flags set to `true` and some custom paths mappings:

```jsonc
"compilerOptions": {
  // ...
  "allowJs": true,
  "checkJs": true,
  "paths": {
    "~/*": [ "./node_modules/*" ],
    "@/*": [ "./src/*" ]
  }
  "baseUrl": "."
  // ...
}
```

The problem is, with `"allowJS": true` TypeScript complains when checking some libraries (in my case, VueJS):
```
[!] (Typescript plugin) TS7006: Parameter 'v' implicitly has an 'any' type.
node_modules\vue\dist\vue.runtime.esm.js (12:19)

12 function isUndef (v) {
                     ~
```
I **need** this config for some other files in my project so I can't simply do without it.

So I tried setting `allowJs` to `false` in `rollup.config.js` for some chunks only:
```js
import ts from '@wessberg/rollup-plugin-ts'

export default {
  // ...
  plugins: [
    ts({
      tsconfig: {
        allowJs: false,
        checkJs: false
      }
    })
  ]
  // ...
}
```
but then the paths mappings are effectively lost:
```
[!] (Typescript plugin) TS2307: Cannot find module '@/common/config'.
src\main\config.ts (6:116)
```



The easy workaround I found is to repeat the path mappings in `rollup.config.js`, like so:
```js
import ts from '@wessberg/rollup-plugin-ts'

export default {
  // ...
  plugins: [
    ts({
      tsconfig: {
        allowJs: false,
        checkJs: false,
        paths: {
            "~/*": [ "./node_modules/*" ],
            "@/*": [ "./src/*" ]
        },
        baseUrl: "."
      }
    })
  ]
  // ...
}
```
However, doing so is not really DRY-friendly. And I loose all other `compilerOptions` from my `tsconfig.json` file. (And I don't want to repeat it entirely neither)

So here's what I finally ended up with:

```js
// rollup.config.js
import ts from '@wessberg/rollup-plugin-ts'
import typescript from 'typescript'

/** Read the project's tsconfig.json and return its compilerOptions */
function getTsconfigCompilerOptions() {

    /** @type {typescript.CompilerOptions} */
    let options = {}

    const parseConfigHost = {
        fileExists: typescript.sys.fileExists,
        readFile: typescript.sys.readFile,
        readDirectory: typescript.sys.readDirectory,
        useCaseSensitiveFileNames: true
    }

    const configFileName = typescript.findConfigFile(
        './',
        typescript.sys.fileExists,
        'tsconfig.json'
    )

    if (configFileName) {
        const configFile = typescript.readConfigFile(configFileName, typescript.sys.readFile)

        options = typescript.parseJsonConfigFileContent(
            configFile.config,
            parseConfigHost,
            './'
        ).options
    }

    return options
}

const tsCompilerOptions = getTsconfigCompilerOptions()

// Later in the file...
export default {
  // ...
  plugins: [
    ts({
      tsconfig: Object.assign({}, tsCompilerOptions, {
        allowJs: false,
        checkJs: false
      })
    })
  ]
  // ...
}

```

The idea is to read the project's `tsconfig.json`, mutate it as needed and then pass the result to `rollup-plugin-ts`.

IMHO this should be the plugin's default behavior. What do you think?

**Comment by wessberg** [comment]
Hi there. Thanks for your detailed comment. I agree that when passing in `CompilerOptions` directly, it should be possible to make whatever is passed act as "overrides" to whatever config can be resolved relative to the current working directory, rather than act as a substitute for resolving a config.

That said, I'm not sure if I think it should be the default. There _are_ use cases for providing `CompilerOptions` directly as plugin options and not expecting them to be merged with anything that might be resolvable on the file system.

What I propose instead is for the `tsconfig` field to also be able to accept a function that receives `CompilerOptions` and enable you to override them. Like:

```typescript
// ...
plugins: [
    ts({
      tsconfig: resolvedConfig => {(...resolvedConfig, allowJs: false, checkJs: false})
    })
]
// ...
```

How do you feel about that approach?

**Comment by Septh** [comment]
Would be awesome!

I also like [rollup-plugin-typescript2](https://github.com/ezolenko/rollup-plugin-typescript2)'s approach where the user can _at the same time_ read their `tsconfig.json` of choice and provide some defaults and some overrides:
```js
 plugins: [
 	typescript({
 		tsconfig: path.join(__dirname, "tsconfig.json"),
 		tsconfigDefaults: defaults,
 		tsconfigOverride: override
 	})
 ]
```
But your approach is much, much cleaner. 👍

**Comment by wessberg** [comment]
I've implemented support for this and released it under v1.1.53.
I've documented usage [in the README](https://github.com/wessberg/rollup-plugin-ts#using-it-with-just-typescript).

But for your convenience, I'm inlining it here:

> You can also pass in a function that receives whatever `CompilerOptions` that could be resolved > relative to the current working directory, but then allow you to override the options:
> 
> ```javascript
> ts({
> 	tsconfig: resolvedConfig => ({...resolvedConfig, allowJs: false})
> });
> ```
> 
> The above example is based on the assumption that a file can be resolved with the name `tsconfig.json`, and if not, the Typescript's default `CompilerOptions` will be used.
> But if you want to provide the name of the `tsconfig` to override, you can also pass in an object > following the following form:
>
> ```javascript
> ts({
> 	tsconfig: {
> 		fileName: "my-awesome-tsconfig.json",
> 		hook: resolvedConfig => ({...resolvedConfig, allowJs: false})
>	}
> });
> ```
